### PR TITLE
torch.compile benchmark utility

### PR DESCRIPTION
**Pull request by msaroufim**
I've had many exchanges that look like this https://github.com/rasbt/faster-pytorch-blog/pull/2 so this is an attempt to get make this problem easier

### CODE REVIEW

File: torch/utils/benchmark/utils/compile.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        else:
+            torchdynamo.reset()
+            compilation_time, running_time = benchmark_compile(model, sample_input, num_iters, backend, None, is_training, optimizer)
+            if running_time is not None:
+                speedup = (eager_time - running_time) / eager_time if eager_time else None
+                table.add_row([("Training" if is_training else "Inference"), backend, "-", compilation_time or "-", running_time, speedup or "-"])
+
+    return table
+
+if __name__ == "__main__":
```
Comment (msaroufim) [inline torch/utils/benchmark/utils/compile.py:168]: This is just convenient for me now locally but will remove everything below this line for merge

File: torch/utils/benchmark/utils/compile.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    
+    class ToyModel(torch.nn.Module):
+        def __init__(self):
+            super(ToyModel, self).__init__()
+            self.weight = torch.nn.Parameter(torch.Tensor(2, 2))
+
+        def forward(self, x):
+            return x * self.weight
+
+    model = ToyModel().cuda()
```
Comment (msaroufim) [inline torch/utils/benchmark/utils/compile.py:179]: this is causing CI to break - I can make a seperate cuda specific and non specific test

File: torch/utils/benchmark/utils/compile.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
+import torch
+
+__all__ = ["bench_all", "benchmark_compile"]
+
+import torch
```
Comment (ezyang) [inline torch/utils/benchmark/utils/compile.py:5]: dupe

File: torch/utils/benchmark/utils/compile.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+import torch._dynamo as torchdynamo
```
Comment (ezyang) [inline torch/utils/benchmark/utils/compile.py:6]: nit: just remove this alias

File: torch/utils/benchmark/utils/compile.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+import time
+import warnings
+from typing import Optional, List
+
+# Added this otherwise all the warnings from each backend will be printed
+# Not sure if this warning will propgate if user is not using this config
+warnings.filterwarnings("ignore")
```
Comment (ezyang) [inline torch/utils/benchmark/utils/compile.py:13]: No, you can't do this. This will clobber whatever user warning setting on import, which is a big no no for libraries. If you really really want to suppress the warnings, do it with a context manager in the utilities (but it's not obvious to me that you want to suppress warnings?)

File: torch/utils/benchmark/utils/compile.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+    assert(inference_table.rows[0][0] == "Inference")
+    assert(inference_table.rows[0][1] == "Eager")
+    assert(inference_table.rows[0][2] == "-")
+    print(inference_table)
+    print("\n===== Training =====")
+    training_table = bench_all(model, torch.ones(1024, 2, 2).cuda(), 5, is_training=True, optimizer=torch.optim.SGD(model.parameters(), lr=0.01))
+    assert(training_table.rows[0][0] == "Training")
+    assert(training_table.rows[0][1] == "Eager")
+    assert(training_table.rows[0][2] == "-")
+    print(training_table)
```
Comment (ezyang) [inline torch/utils/benchmark/utils/compile.py:192]: fix missing trailing newline

File: torch/utils/benchmark/utils/compile.py (hunk 3 above)
Comment (ezyang) [reply torch/utils/benchmark/utils/compile.py:5]: nit: can you opt this file into UFMT? Thanks

File: torch/utils/benchmark/utils/compile.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+        header = " | ".join([f"{str(field_name).ljust(width)}" for field_name, width in zip(self.field_names, column_widths)])
+        separator = "-+-".join(["-" * width for width in column_widths])
+
+        rows = []
+        for row in formatted_rows:
+            rows.append(" | ".join([f"{cell.ljust(width)}" for cell, width in zip(row, column_widths)]))
+
+        return f"{header}\n{separator}\n" + "\n".join(rows)
+
+def bench_loop(model : torch.nn.Module, sample_input : torch.Tensor, num_iters : int, is_training : bool =False, optimizer : torch.optim.Optimizer = None):
```
Comment (ezyang) [inline torch/utils/benchmark/utils/compile.py:57]: Some of these want to be kwargs only, probably

File: torch/utils/benchmark/utils/compile.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+    It also takes care of calling `cuda.synchronize()` if the model is on GPU
+    """
+    durations = []
+    for _ in range(num_iters):
+        start = time.time()
+        
+        if is_training and optimizer:
+            optimizer.zero_grad()
+            output = model(sample_input)
+            loss = output.sum()
```
Comment (ezyang) [inline torch/utils/benchmark/utils/compile.py:70]: This doesn't always work; e.g., maskrcnn returns a dictionary of stuff

File: torch/utils/benchmark/utils/compile.py [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    """
+    This is a simple loop that can be used to benchmark a model for either training or inference
+    It takes care of taking several measurements and averaging them
+    It also takes care of calling `cuda.synchronize()` if the model is on GPU
+    """
+    durations = []
+    for _ in range(num_iters):
+        start = time.time()
+        
+        if is_training and optimizer:
```
Comment (ezyang) [inline torch/utils/benchmark/utils/compile.py:67]: is_training does nothing unless you pass optimizer; in that case, is there any point in an is_training parameter?

File: torch/utils/benchmark/utils/compile.py [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+            loss = output.sum()
+            loss.backward()
+            optimizer.step()
+        else:
+            model(sample_input)
+        
+        end = time.time()
+
+        if sample_input.get_device() >= 0:
+            torch.cuda.synchronize()
```
Comment (ezyang) [inline torch/utils/benchmark/utils/compile.py:79]: This seems wrong? Don't you have to synchronize before you finish timing?

**Comment by ezyang** [comment]
Given that the inner benchmarking loop here also wasn't done correctly, I'd like to suggest that we figure out how to use the pre-existing benchmarking helpers as the inner loop, and then have the "iterate through all backends" as something built on top of it.

File: torch/utils/benchmark/utils/compile.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+    It takes care of taking several measurements and averaging them
+    It also takes care of calling `cuda.synchronize()` if the model is on GPU
+    """
+    durations = []
+    for _ in range(num_iters):
+        start = time.time()
+        
+        if is_training and optimizer:
+            optimizer.zero_grad()
+            output = model(sample_input)
```
Comment (ngimel) [inline torch/utils/benchmark/utils/compile.py:69]: assuming that sample input is always a single tensor is restrictive?

File: torch/utils/benchmark/utils/compile.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+
+    The important warnings are
+    Your model is loaded on GPU
+    Your GPU supports tensor cores, we will enable it automatically by setting `torch.set_float32_matmul_precision('high')`
+
+    If a compilation fails for any reason including the dependency not being included then we will print Failed to compile {backend} with mode {mode}
+
+
+    """
+    if next(model.parameters()).is_cuda:
```
Comment (ngimel) [inline torch/utils/benchmark/utils/compile.py:131]: there are models without parameters

File: torch/utils/benchmark/utils/compile.py [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+        print("Your model is loaded on GPU")
+        if torch.backends.cuda.matmul.allow_tf32 is False and torch.cuda.get_device_capability() >= (8, 0):
+            print("Your GPU supports tensor cores, we will enable it automatically by setting `torch.set_float32_matmul_precision('high')`")
+            torch.set_float32_matmul_precision("high")
```
Comment (ngimel) [inline torch/utils/benchmark/utils/compile.py:135]: you never unset it, it's better to do these things in `try...finally` block so that even compilation fails or there's some other error, the initial state is restored.

File: torch/utils/benchmark/utils/compile.py (hunk 8 above)
Comment (msaroufim) [reply torch/utils/benchmark/utils/compile.py]: Will make it possible for people to add a custom loss function otherwise default to this

File: torch/utils/benchmark/utils/compile.py [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+
+# Added this otherwise all the warnings from each backend will be printed
+# Not sure if this warning will propgate if user is not using this config
+warnings.filterwarnings("ignore")
+
+
+
+class SimpleTable:
+    """
+    This is a simple table class that can be used to print a table of data.
```
[the remark this answers is not in the document]
Comment (msaroufim) [reply torch/utils/benchmark/utils/compile.py]: Yup it worked just fine thanks for pointing that out

[the remark this answers is not in the document]
Comment (msaroufim) [reply torch/utils/benchmark/utils/compile.py]: Granted it seems like it's not installed on a bunch of CI instances but can still make it work

**Comment by msaroufim** [comment]
> Given that the inner benchmarking loop here also wasn't done correctly, I'd like to suggest that we figure out how to use the pre-existing benchmarking helpers as the inner loop

Problems should be fixed, I unfortunately couldn't find a helpful utility that did what I wanted

**Comment by ngimel** [comment]
`torch.utils.benchmark.Timer.blocked_autorange` or `torch.utils.benchmark.Timer.timeit` should work?

File: torch/utils/benchmark/utils/compile.py [hunk 15]
```diff
@@ -34,36 +35,37 @@ def _disable_tensor_cores():
...
+    optimizer.zero_grad()
+            """
+        else:
+            # Inference mode
+            stmt = "model(sample_input)"
+
+        # Create the Timer object
+        timer = Timer(
+            stmt=stmt,
+            globals={"model": model, "sample_input": sample_input, "optimizer": optimizer, "loss_fn": loss_fn}
```
Comment (ngimel) [inline torch/utils/benchmark/utils/compile.py:59]: loss_fn and optimizers are not necessarily assigned?

Comment (msaroufim) [reply torch/utils/benchmark/utils/compile.py:59]: I might be missing something but they both default to None values, would you prefer if I explicitly checked if they were None before constructing the globals dict?

Comment (ngimel) [reply torch/utils/benchmark/utils/compile.py:59]: You are right, that's fine

File: torch/utils/benchmark/utils/compile.py [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+    loss.backward()
+    optimizer.step()
+    optimizer.zero_grad()
+            """
+        else:
+            # Inference mode
+            stmt = "model(sample_input)"
+
+        # Create the Timer object
+        timer = Timer(
```
Comment (msaroufim) [inline torch/utils/benchmark/utils/compile.py:57]: Ed gave us some interesting feedback on this, the compilation time is eaten every time so we have a few workarounds
1. Add a warmup to timer object so the compilation is ignored but only for when we're not explicitly measuring compilations
2. quiescence mode via torch._dynamo.run can help us pass the dynamo cache as a global variable instead
3. In tests we can use CompileCounter to explicitly make sure that no new compilation is happening

File: test/test_compile_benchmark_util.py [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+                self.weight = torch.nn.Parameter(torch.Tensor(2, 2))
+
+            def forward(self, x):
+                return x * self.weight
+
+        torchdynamo.reset()
+        model = ToyModel().cuda()
+
+        inference_table = bench_all(model, torch.ones(1024, 2, 2).cuda(), 5)
+        self.assertTrue("Inference" in inference_table and "Eager" in inference_table and "-" in inference_table)
```
Comment (ezyang) [inline test/test_compile_benchmark_util.py:36]: If you want to test the printing, one way might be to separate data collection and data printing; then the data printing can be tested deterministically with an expect test and synthetic data.

File: torch/utils/benchmark/utils/compile.py [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+import torch._dynamo
+from torch.utils.benchmark import Timer
+
+from typing import Optional, List, Callable, Union, Any, cast
+
+try:
+    from tabulate import tabulate
+    HAS_TABULATE = True
+except ImportError:
+    HAS_TABULATE = False
```
Comment (ezyang) [inline torch/utils/benchmark/utils/compile.py:14]: nit: Let's just require tabulate to be installed for simplicity, it's not a big deal

File: test/test_compile_benchmark_util.py (hunk 17 above)
Comment (msaroufim) [reply test/test_compile_benchmark_util.py:36]: Can the compilation times and running times ever be deterministic though?

File: torch/utils/benchmark/utils/compile.py (hunk 18 above)
Comment (msaroufim) [reply torch/utils/benchmark/utils/compile.py:14]: Alright I'll just add it to requirements.txt

Comment (ezyang) [reply torch/utils/benchmark/utils/compile.py:14]: You don't have to add it to requirements.txt, the idea is just that to use this file, you'll need tabulate

File: torch/utils/benchmark/utils/compile.py [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+_default_float_32_precision = torch.get_float32_matmul_precision()
+
+
+if HAS_TABULATE:
+    def _enable_tensor_cores():
+        global _warned_tensor_cores
+
+        if torch.cuda.is_available():
+            if torch.backends.cuda.matmul.allow_tf32 is False and torch.cuda.get_device_capability() >= (8, 0):
+                torch.set_float32_matmul_precision("high")
```
Comment (ezyang) [inline torch/utils/benchmark/utils/compile.py:26]: I have a noob question: why is the effect of tensor cores not proportional across all backends? E.g., if I'm comparing eager and inductor, as long as I compare them consistently with or without tensor cores, I feel like I should get the same factor? Does tensorcores change where the bottlenecks are?

Comment (ezyang) [reply torch/utils/benchmark/utils/compile.py:26]: If the end user isn't using tensor cores, it seems disingenous to give benchmark numbers with tensor cores, and then they will proceed to not use it.

**Comment by ezyang** [comment]
Can you add some sample output to the PR desc?

File: torch/utils/benchmark/utils/compile.py [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+                torch._dynamo.reset()
+                compilation_time, running_time = benchmark_compile(
+                    model, sample_input, num_iters, backend, None, optimizer, loss_fn)
+
+                if running_time is not None:
+                    speedup = eager_time / running_time
+                    table.append([
+                        ("Training" if optimizer else "Inference"),
+                        backend, "-", compilation_time or "-", running_time, speedup or "-"
+                    ])
```
Comment (ezyang) [inline torch/utils/benchmark/utils/compile.py:172]: Can you check that the units on each table entry print the same?

File: torch/utils/benchmark/utils/compile.py [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+        """
+        if backend:
+            try:
+                torch._dynamo.reset()
+                opt_model = torch.compile(model, backend=backend, mode=mode)
+
+                # Compilation only happens after the first inference
+                compilation_time = bench_loop(opt_model, sample_input, 1, optimizer, loss_fn)
+
+            except BaseException as e:
```
Comment (ezyang) [inline torch/utils/benchmark/utils/compile.py:90]: How come BaseException?

File: torch/utils/benchmark/utils/compile.py (hunk 16 above)
Comment (ezyang) [reply torch/utils/benchmark/utils/compile.py:57]: So for this tool, I think we should (1) measure how long it takes to compile and report that, but (2) error if we end up recompiling on subsequent iteration loops

File: torch/utils/benchmark/utils/compile.py (hunk 19 above)
Comment (ngimel) [reply torch/utils/benchmark/utils/compile.py:26]: Tensorcores move bottlenecks from matmul (that inductor doesn't help with) to other, bw-bound, ops that inductor fuses. E.g. w/o tensorcores 90% of eager workload can be in matmuls, so if inductor accelerates remaining 10% by 2x, it's 5% e2e, with tensorcores matmuls would be say 60%, so accelerating remainder by 2x is 20% e2e.

Comment (ezyang) [reply torch/utils/benchmark/utils/compile.py:26]: alright, thanks

File: torch/utils/benchmark/utils/compile.py (hunk 18 above)
Comment (msaroufim) [reply torch/utils/benchmark/utils/compile.py:14]: Ah gotcha so still skip test in CI

**Comment by msaroufim** [comment]
> Can you add some sample output to the PR desc?

The table is too wide and linter is not happy, lemme see if I can reduce the number of significant bits to make this work

> How come base exception
Linter doesnt like naked exceptions and suggests base, there's also different reasons why a backend might fail, missing package or any arbitrary exception triggered from that backend codebase

>Can you check that the units on each table entry print the same?
Yes makes a lot sense will do

**Comment by ezyang** [comment]
> Linter doesnt like naked exceptions and suggests base, there's also different reasons why a backend might fail, missing package or any arbitrary exception triggered from that backend codebase

BaseException includes KeyboardInterrupt. You probably don't want to catch that. Catch Exception instead.

File: torch/utils/benchmark/utils/compile.py [hunk 22]
```diff
@@ -134,19 +134,19 @@ def bench_all(
...
-        | Train/Inference   | Backend         | Mode   | Compilation Time   | Average Running Time   | Speedup   |
-        |-------------------|-----------------|--------|--------------------|------------------------|-----------|
-        | Inference         | Eager           | -      | -                  | 0.0163 ms              | -         |
-        | Inference         | aot_ts_nvfuser  | -      | 0.1914 ms          | 0.088 ms               | 19%       |
-        | Inference         | cudagraphs      | -      | 0.2405 ms          | 0.1286 ms              | 13%       |
-        | Inference         | nvprims_nvfuser | -      | 0.1898 ms          | 0.1312 ms              | 12%       |
+        | Train/Inference   | Backend         | Mode   | Compilation Time   | Average Running Time   | Speedup        |
+        |-------------------|-----------------|--------|--------------------|------------------------|----------------|
+        | Inference         | Eager           | -      | -                  | 0.0172 ms              | -              |
+        | Inference         | aot_ts_nvfuser  | -      | 0.1658 ms          | 0.0789 ms              | 0.2180x slower |
```
Comment (ezyang) [inline torch/utils/benchmark/utils/compile.py:140]: nit: four digits is probably too many sigdigs, probably two is enough.

**Review by ezyang — APPROVED** [review]
ok, in the interest of moving this along
